librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    yin_denominator = cumulative_mean[min_period-2:max_period, :]
+    yin_frames = yin_numerator / (yin_denominator + util.tiny(yin_denominator))
+
+    # Parabolic interpolation.
+    parabola_a = (yin_frames[:-2, :] + yin_frames[2:, :] - 2*yin_frames[1:-1, :]) / 2
+    parabola_b = (yin_frames[2:, :] - yin_frames[:-2, :]) / 2
+    parabolic_shifts = -parabola_b / (2*parabola_a + util.tiny(parabola_a))
+    parabolic_shifts[np.abs(parabolic_shifts) > 1] = 0
+    yin_frames = yin_frames[1:-1, :]
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    n_thresholds = 100  # number of thresholds for peak estimation.
+    thresholds = np.linspace(0, 1, n_thresholds + 1)
+    beta_cdf = scipy.stats.beta.cdf(thresholds, 2, 18)
+    beta_probs = np.diff(beta_cdf)
+
+    yin_probs = np.zeros_like(yin_frames)
+    for i, yin_frame in enumerate(yin_frames.T):
+        # 2. For each frame find the troughs.
+        is_trough = np.logical_and(
+            yin_frame[1:-1] < yin_frame[:-2],
+            yin_frame[1:-1] < yin_frame[2:])
+        trough_index, = np.nonzero(is_trough)
+
+        if len(trough_index) == 0:
+            continue
+
+        # 3. Find the troughs below each threshold.
+        trough_heights = yin_frame[trough_index + 1]
+        trough_thresholds = trough_heights[:, None] < thresholds[None, 1:]
+
+        # 4. Define the prior over the troughs.
+        # Smaller periods are weighted more.
+        trough_positions = np.cumsum(trough_thresholds, axis=0) - 1
+        n_troughs = np.count_nonzero(trough_thresholds, axis=0)
+        trough_prior = scipy.stats.boltzmann.pmf(trough_positions, 10, n_troughs)
+        trough_prior[~trough_thresholds] = 0
+
+        # 5. For each threshold add probability to global minimum if no trough is below threshold,
+        # else add probability to each trough below threshold biased by prior.
+        probs = np.sum(trough_prior * beta_probs, axis=1)
+        global_min = np.argmin(trough_heights)
+        n_thresholds_below_min = np.count_nonzero(~trough_thresholds[global_min, :])
+        probs[global_min] += 0.01 * np.sum(beta_probs[:n_thresholds_below_min])
+
+        yin_probs[trough_index + 1, i] = probs
+
+    yin_period, frame_index = np.nonzero(yin_probs)
+
+    # Refine peak by parabolic interpolation.
+    period_candidates = min_period + yin_period
+    period_candidates = period_candidates + parabolic_shifts[yin_period, frame_index]
+    f0_candidates = sr / period_candidates
+
+    n_bins_per_semitone = 10  # bins for discretizing the pitch space between fmin and fmax.
+    n_pitch_bins = int(np.log2(fmax / fmin) * 12 * n_bins_per_semitone)
+
+    # Construct transition matrix.
+    transition_width = 10 * (n_bins_per_semitone // 2) + 1
```

Comment:
why 10?

Reply:
This parameter comes from the pYIN paper:
>The latter is realised as a triangular weight distribution which encodes that a pitch jump can be at most 25 ‘bins’, which corresponds to 2.5 semitones per frame.

In the paper, they use a `hop_length` of 5.8ms so we might need to change the value for best performance. Do you have any suggestions?

Reply:
I see a few options here.

1. Translate 2.5semitones/frame to our default settings, which I guess would be 10 semitones/frame (with hop ~= 23ms), so `10 -> 40`
2. Allow the user to specify the number of bins to jump per frame
3. Allow the user to specify a maximum chirp rate in natural units (probably in octaves/sec, to be invariant to the number of bin subdivisions) and do the math.  The default settings would be 2.5 semitones / frame, 4.8 frames to get an octave, so ~0.0278.

I think I'm leaning toward option (3), since it does the best to decouple the various parameter interactions while maintaining prior behavior.  What do you think?

Reply:
I've been trying a few of the options. I think the problem with option 3 is that it's not very intuitive to set a good value. Correct me if I'm wrong but shouldn't 2.5 semitones/frame at a hop of 5.8ms translate to `2.5 / 12 / 0.0058 = 35.91` octaves/sec. I think I'd find it difficult to interpret this without converting it to semitones per frame.

Reply:
> I think the problem with option 3 is that it's not very intuitive to set a good value.

That's a fair point.  Ideally, users shouldn't have to muck with this too much if we get a reasonably good default setting.



>  Correct me if I'm wrong but shouldn't 2.5 semitones/frame at a hop of 5.8ms translate to `2.5 / 12 / 0.0058 = 35.91` octaves/sec.

Yeah, I had calculated seconds/octave instead. Oops!

> I think I'd find it difficult to interpret this without converting it to semitones per frame.

I generally agree, but the reason to do it in octaves/sec is that if a user changes the frame rate, then the behavior should be (relatively) stable.  If you parametrize it by semitones/frame, and the user changes the frame rate, then they also have to know how to change the semitones/frame parameter.  That puts more work on the user, and I'd like to avoid that if possible.